### PR TITLE
Sync server

### LINKED ISSUE #18: Add Coveralls Go coverage + badge to repo
what it says

### LINKED ISSUE #20: migrate Sync Server to GCP / Go

### LINKED ISSUE #24: Test new Sync Server output

### PR DESCRIPTION
**Pull request by rheinardkorf**
### Todo
- [x] Create a Sync Server
- [x] Verify parity
- [x] Setup local development
- [x] Test all three SQS queues
- [x] Update stdout to only include warnings/errors/failures to remove the noise when the sync is happening.
- [x] Finalize GCP cluster settings and look into running the Sync Server as a single container that has a persistent storage disk attached to store the DB.
- [x] Code review

closes #18, closes #20, closes #24

### CODE REVIEW

**Comment by rheinardkorf** [comment]
@valendesigns I have opened the PR, but it will not be ready until we have added the containerisation.

**Comment by valendesigns** [comment]
@rheinardkorf Tomorrow, could you troubleshoot why the sync server doesn't work when you bring the container up.

**Comment by rheinardkorf** [comment]
@valendesigns it was the SSL root certs again.

I've added a command to the Makefile `sync.update.certs`.  It gets it from the `curl` mirror, but they request it not being done more than once a day, so I'm not including it in the build step.  Each time we're ready for deployment would be a good time to get new root certs, commit it to the repo, and rebuilt the image.

This gets the sync-server running for me now in the container.  Please review the changes.

**Comment by valendesigns** [comment]
@rheinardkorf How do the latest changes look to you? I'm still testing a bit locally, but should be doing a more thorough review of the Go code tomorrow.

File: cmd/sync-server/main.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+		Queues: map[string]struct {
+			Region   string
+			Key      string
+			Secret   string
+			Queue    string
+			Endpoint string
+			Active   bool
+			Accepts  string // "all" or "themes" or "plugins"
+		}{
+			"primary": {
```
Comment (valendesigns) [inline cmd/sync-server/main.go]: I'm thinking that we should have two queues, a PHPCS queue and a Lighthouse queue. What purpose does this configuration serve? We don't have an immediate need for a plugin queue and being explicit about what the queue is makes more sense semantically to me than generic queues for themes and plugins.

File: cmd/sync-server/main.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	}
+}
+
+// infoWorker reads a project from the projects channel, runs it through the update check
+// and (conditionally) sends it to the dispatcher which loads up the job queue.
+func infoWorker(projects <-chan wporg.RepoProject, checker sync.UpdateChecker, dispatcher sync.Dispatcher, messages chan *message.Message) {
+	for {
+		select {
+		case project := <-projects:
+			// If this project is not in sync (or forced), then sent it to the queue.
```
Comment (valendesigns) [inline cmd/sync-server/main.go]: There is a typo in the comment `sent` should be `send`.

File: cmd/sync-server/main.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+			}
+		}
+	}
+}
+
+// defaultStandards gets the standards to run against a project.
+func defaultStandards() []string {
+	return []string{
+		"phpcs_wordpress",
+		"phpcs_phpcompatibility",
```
Comment (valendesigns) [inline cmd/sync-server/main.go:182]: What about lighthouse?

File: cmd/sync-server/main.go (hunk 2 above)
Comment (valendesigns) [reply cmd/sync-server/main.go]: fixed in [#610f777](https://github.com/xwp/go-tide/pull/23/commits/610f777988873bc6d2fb32688bd9c8dbfafdcc14)

File: cmd/sync-server/main.go (hunk 3 above)
Comment (rheinardkorf) [reply cmd/sync-server/main.go:182]: we can add that, but it should also be added below to `defaultAudits`...  once we sort out the audit server I'd like to review this.

Comment (valendesigns) [reply cmd/sync-server/main.go:182]: Does Lighthouse have any real options being passed to it?

Comment (valendesigns) [reply cmd/sync-server/main.go:182]: We'll double back on this once we've identified what the options would be.

**Review by valendesigns — APPROVED** [review]

**Pull request merged**